### DISCUSSION TITLE
Login with email/phone number and otp login

### DISCUSSION THREAD

**Discussion by kumararvindibs** [keycloak/keycloak#19167] [Q&A]
HI Team,

I'm very new to the keycloak cofiguration. 
Is there any way to log in with a mobile number or email id and OTP (not password)?
Please suggest. 
Thanks in Advance

**Comment by darius-m** [comment]
What do you mean by "mobile number or email"? Using a OTP token sent by SMS or email or a magic link? If so, as far as I know Keycloak does not support either of them, but there may be some custom SPIs that you could use. If you simply want to authenticate using an OTP generated by an authenticator application, that should be pretty easy if you create an authentication flow that contains an username form and then requires an OTP code.

**Reply by kumararvindibs** [reply]
Thanks for your reply.

I want to logged-in user by providing the mail id or mobile number then an OTP will be sent to the mail id or mobile number respectively. The user will enter the OTP then after the validation of the OTP it should be logged in to the keycloak. Although I have no idea till now how it will be logged in after providing the valid OTP, this is also an open question for me.
The Keycloak expects the username and password and emailId and password so that I also need to take care.

Using the authenticator app its pretty simple I have tried but my requirement is different.

I also tried to add the Authentication flow but it allows us only the predefined steps which don't have such mechanism.

**Reply by ismailhozza** [reply]
Hi @kumararvindibs I have a similar requirement. Have you found a solution to implement the OTP based authentication?

**Comment by kumararvindibs** [comment]
HI Hussein,

I didn't find the solution for this approach from keycloak side. So I'm
using mail server to send the otp and verifying it by custom code. This
approach is working for me. As of now I'm using email otp not for mobile.

Regards,
Arvind Kumar

**Reply by ismailhozza** [reply]
Thanks!

**Comment by codespearhead** [comment]
Can you please close this issue/discussion [1][2]?

This question has been asked several times before. You can find previous discussions by searching for “sms” or “email” in the GitHub search bar of the Keycloak repository [3][4].

Currently, there is no official support for using email or phone numbers for authentication in Keycloak. Therefore, any issues you encounter with third-party solutions should be reported in their respective repositories, not here.

If you're interested in this feature, feel free to share your use case here [5], and upvote it by adding a ⬆️ to the discussion’s description after reading this answer carefully [6].

***

[1] https://docs.github.com/en/issues/tracking-your-work-with-issues/administering-issues/closing-an-issue
[2] https://docs.github.com/en/discussions/managing-discussions-for-your-community/managing-discussions#closing-a-discussion
[3] https://github.com/search?q=repo%3Akeycloak%2Fkeycloak+%22SMS%22&type=issues
[4] https://github.com/search?q=repo%3Akeycloak%2Fkeycloak+%22SMS%22&type=discussions
[5] https://github.com/keycloak/keycloak/discussions/27152
[6] https://github.com/keycloak/keycloak/discussions/27152#discussioncomment-10923758
